flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java

```diff
@@ -89,7 +91,78 @@ public void testInvalidVersionRejection() throws Exception {
 		} catch (IllegalArgumentException e) {
 			// expected
 		}
+	}
 
+	/**
+	 * Tests that we fail the operation if the remote connection closes.
+	 */
+	@Test
+	public void testConnectionClosedHandling() throws Exception {
+		final ServerSocket serverSocket = new ServerSocket(0);
+		final String targetAddress = "localhost";
+		final int targetPort = serverSocket.getLocalPort();
+
+		try (final RestClient restClient = new RestClient(RestClientConfiguration.fromConfiguration(new Configuration()), TestingUtils.defaultExecutor())) {
+			final CompletableFuture<EmptyResponseBody> responseFuture = restClient.sendRequest(
+				targetAddress,
+				targetPort,
+				new TestMessageHeaders(),
+				EmptyMessageParameters.getInstance(),
+				EmptyRequestBody.getInstance(),
+				Collections.emptyList());
+
+			// establish connection
+			final Socket connectionSocket = serverSocket.accept();
+
+			// close connection
+			connectionSocket.close();
+
+			try {
+				responseFuture.get();
+			} catch (ExecutionException ee) {
+				assertThat(ExceptionUtils.findThrowable(ee, ConnectionClosedException.class).isPresent(), is(true));
+			}
+		} finally {
+			serverSocket.close();
+		}
+	}
+
+	/**
+	 * Tests that we fail the operation if the client closes.
+	 */
+	@Test
+	public void testRestClientClosedHandling() throws Exception {
+		final ServerSocket serverSocket = new ServerSocket(0);
+		final String targetAddress = "localhost";
+		final int targetPort = serverSocket.getLocalPort();
+		Socket connectionSocket = null;
+
+		try (final RestClient restClient = new RestClient(RestClientConfiguration.fromConfiguration(new Configuration()), TestingUtils.defaultExecutor())) {
+			final CompletableFuture<EmptyResponseBody> responseFuture = restClient.sendRequest(
```

Comment:
I'm wondering...consider the `submitRequest` method from the `RestClient`:
```
private <P extends ResponseBody> CompletableFuture<P> submitRequest(String targetAddress, int targetPort, Request httpRequest, JavaType responseType) {
		final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);

		final CompletableFuture<Channel> channelFuture = new CompletableFuture<>();

		connectFuture.addListener(
			(ChannelFuture future) -> {
				if (future.isSuccess()) {
					channelFuture.complete(future.channel());
				} else {
					channelFuture.completeExceptionally(future.cause());
				}
			});

		return channelFuture
			.thenComposeAsync(
				channel -> {
					ClientHandler handler = channel.pipeline().get(ClientHandler.class);
					CompletableFuture<JsonResponse> future = handler.getJsonFuture();
					try {
						httpRequest.writeTo(channel);
					} catch (IOException e) {
						return FutureUtils.completedExceptionally(new FlinkException("Could not write request.", e));
					}
					return future;
				},
				executor)
			.thenComposeAsync(
				(JsonResponse rawResponse) -> parseResponse(rawResponse, responseType),
executor);
```

What happens if the client is shutdown, before the `thenComposeAsync` portions have been run? Then the channelFuture is completed by the listener; and then subsequent stages aren't executed?

Reply:
I think this situation is fine, because an exceptionally completed future won't trigger the executions of the async stages. Instead it will directly fail the returned future.

Reply:
but it does fail with a different exception, no? although from a user-perspective the same issue happened.